### ISSUE TITLE
Procedure for evolution of language spec

### ISSUE THREAD

**Issue by AnastasiaStulova** [KhronosGroup/OpenCL-Docs#65]
Even though the specs are published some time ago it is desirable to have the ability to improve them and fix the bugs.

The question is how to change existing specs and propagate changes to tools and testing (in particular CTS) minimizing negative impact.

Summary of previous discussions:
- If changes affect the diagnostics i.e. some code would be rejected or accepted differently from the older specs. What should happen to CTS and how to dial with the drivers that are already shipped that will behave differently to the new drivers?
- Adding an extension for each modification in a language doesn't seem like a great language design choice.
- Changing language version for a group of features could be an option  but (a) how to define the amount of changes that is enough (b) how long would it take to wait for a new version release.
- Clang extension is not portable and has some limitations i.e. ideally it shouldn't be big changes.

Some early observations:
- Clang has one language mode for each standard i.e. -std=c++11. It documents revision that it implements and deltas between releases however there is no formal procedure for this.
- TODO: Check the procedure with ISO.

**Comment by AnastasiaStulova** [comment]
One idea could be to use minor version and let compiler add implicit macro indicating which minor version it implements for a certain standard.

Example: If we are to publish changes to CL2.0

We could add a macro `__CL_VERSION_MINOR__` that could have values
`CL_MINOR_VERSION_CL_2_0_0` for original spec
`CL_MINOR_VERSION_CL_2_0_1` for the spec with the changes to the original revision
`CL_MINOR_VERSION_CL_2_0_2` ...

This will allow to query what compiler supports and condition the code to make sure it's portable between drivers if needed.

We would need to define the process for publishing the minor spec versions. The CTS can use the defined implicit macros to condition the code too. It is relatively easy compiler change because it would still need to support one language mode. However, we would need to think of the process for the macro to be available to the users (is it at the beginning of the new version development phase or at the end).

**Comment by keryell** [comment]
Perhaps we could use something similar to C++20 feature testing macros https://en.cppreference.com/w/cpp/feature_test ?

**Comment by davidrohr** [comment]
I agree that it should be transparent to the programmer as long as it is related to minor revisions / bug fixes. For real bug-fixes, I am not even sure how prominently I would present them, in particular when they are backward compatible. In particular general minor improvements should not become OpenCL extensions. This would only complicate things in my opinion. Extension make most sense for something that might not be generally available also in the future.

I am not sure if, as a programmer, I want to check for all required features and implement workarounds if they are not present. Ideally, I would just assume they are available, and fail otherwise.

For new features, I think both a minor version or feature testing macros are fine. In general, I like the macros a bit more, since they simplify the process of testing one particular feature. One should make sure, however, to not overdo it with too many macros.

**Comment by AnastasiaStulova** [comment]
Proposed way forward:
**1.** Allow language extensions for the changes that don't require any update in the API or SPIR-V format.
**2.** Language extensions can be documented in a stand alone document against certain language versions.
**3.** A mechanism to query an extension is to be provided in the format of "__" prefixed predefined macros.

_Question_: would we have unique macro available for every extension or would we use macro versions to indicate different extensions of similar type?

_Example_: `__cpp_constexpr` has values `200704L`, `201304L`, `201603L` depending on the C++ standard that compiler supports. I can't think of any example where we would find it useful right now but this is a more generic mechanism as soon as we find some meaningful versioning scheme.

**4.** Language extensions can migrate into a language standard under the new version at some point of time. Once this happens checking the extension presence will no longer be required if compiling with that standard version.
